### ISSUE TITLE
ops.linemerge fails when high precision required

### ISSUE THREAD

**Issue by philippschw** [Toblerity/Shapely#457]
I noticed that linemerge fails when the linestrings have many digits behind the comma. Alhough, it should return linestring from a  multilinestring if nodes are equal it keeps returning a multilinestring.  I read there is an ongoing discussion to develop  a precision model for shapely, would this help to resolve the problem for this case? 
What I actually want to do is connect contiguous Line Features. Hence I perform union followed by linemerge.
For geopandas and knowning that this linestrings are contiguous: `linemerge(gdf.loc[4].geometry.union(gdf.loc[0].geometry))`
Could you suggest a (temporary) solution for my problem? 
I realized that ops.snap works with a very small tolerance works fairly well, but struggles also with the same precision problem.

Simple Example to illustrate

```
lines = MultiLineString([ (  (88.41475335106594, 26.633113517780163),
                             (88.41476094453927, 26.633109647037365),
                             (88.41477625337141, 26.633092870351657)),
                         
                             ((88.41477625337141, 26.633092870351657),
                             (88.41479118130584, 26.633104328661116),
                             (88.41478356498017, 26.633109893505278),
                             (88.41477625337141, 26.633092870351657))
                        ])
type(linemerge(lines))
```

> shapely.geometry.multilinestring.MultiLineString

```
lines = MultiLineString([ (  (88.41475, 26.63311),
                             (88.41476, 26.63310),
                             (88.41477, 26.63309)),
                         
                             ((88.41477, 26.63309),
                             (88.414791, 26.633104),
                             (88.414783, 26.633109),
                             (88.414776, 26.633092))
                        ])

type(linemerge(lines))
```

> shapely.geometry.linestring.LineString

**Comment by sgillies** [comment]
@philippschw a future fixed-precision model might help, but I think the issue in the first half of your example is that the second part of the MultiLineString is virtually closed, i.e. the first and last point are the same, while this is not the case in the second half of your example (`(88.41477, 26.63309)` is significantly different from `(88.414776, 26.633092)`. Try this, where I've changed just the last coordinate in your example to close the second part of the MultiLineString:

```python
>>> from shapely.geometry import *
>>> lines = MultiLineString([ (  (88.41475, 26.63311),
...                              (88.41476, 26.63310),
...                              (88.41477, 26.63309)),
...
...                              ((88.41477, 26.63309),
...                              (88.414791, 26.633104),
...                              (88.414783, 26.633109),
...                              (88.41477, 26.63309))
...                         ])
>>> from shapely.ops import linemerge
>>> linemerge(lines)
<shapely.geometry.multilinestring.MultiLineString object at 0x10140c710>
```

I think the linemerge algorithm simply does not merge closed lines with other lines.

**Comment by philippschw** [comment]
@sgillies thank you for your response. I get the same result as you with the example you provide. The coordinates I presented are coming from a raw shapefile of raw data. I assume the underlying algorithm can or should not be changed. Do you have any suggestions, how I could clean the LineString raw data, to avoid this problem?

**Comment by sgillies** [comment]
I've made a little sketch to illustrate what I think is at the heart of the problem.

![20170116_143136](https://cloud.githubusercontent.com/assets/33697/21984985/f9d0736e-dbf8-11e6-91ae-04b4e9d6d10d.jpg)

When you have a closed line (part 1, with the point shared by its ends shown as an open circle) and another line (part 0) with an end that connects to that closing point, I think the algorithm breaks down. Part 0 could merge to either end of 1, right? Instead of guessing which one you want, it refuses to merge. In this way, GEOS follows the [Zen of Python](https://www.python.org/dev/peps/pep-0020/#the-zen-of-python).

> In the face of ambiguity, refuse the temptation to guess.

If you know which ends of parts 0 and 1 *should* be merged, move the other end point of 1 a bit and then move it back. That's all I've got, sorry. It could be worth taking your question to GIS StackExchange to see if any other GEOS or PostGIS experts have tricks to share.
